File: python/day08.py
import util
import math
NAME = "Camel Cards"

input = util.get_input("day08")

example = """
LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)""".strip().splitlines()

MOVEDIR = { "L":0, "R":1}

def parse_network(lines):
    return  {n[:3]:(n[7:10], n[12:15]) for n in lines}

part1_answer = 23147
def part01(data):
    instructions, network = (data[0], parse_network(data[2:]))
    print(network)
    pos = "AAA"
    n = 0
    while pos != "ZZZ":
        mov = MOVEDIR[instructions[n % len(instructions)]]
        pos = network[pos][mov]
        n+=1
    print(n)

def get_first(p, instr, nw):
    n = 0
    while p[-1] != "Z":
        ip = n % len(instr)
        p = nw[p][MOVEDIR[instr[ip]]]
        n+=1 
    return n,p

# part2_answer = 48132530481716685194553863 too high
part2_answer = 22289513667691
def part02(data):
    instructions, network = (data[0], parse_network(data[2:]))
    pos = [k for k in network.keys() if k[-1] == "A"] 
    firstdists = [get_first(p, instructions, network)[0] for p in pos]
    gcd = math.gcd(*firstdists)
    print(gcd * math.prod([int(d/gcd) for d in firstdists]))

File: rust/src/day08.rs
use std::collections::HashMap;

use num::integer::lcm;

use crate::util;

pub const NAME: &str = "Day 8: Haunted Wasteland";

pub fn _example() -> String {
    String::from(
        "LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)",
    )
}

pub fn data() -> String {
    util::get_input("day08.txt")
}

#[cfg(test)]
mod test_result {
    use super::*;
    #[test]
    fn part01() {
        assert_eq!(super::part01(data()), 0);
    }
    #[test]
    fn part02() {
        assert_eq!(super::part02(data()), 0);
    }
}
#[cfg(test)]
mod test_example {
    use super::*;
    #[test]
    fn part01() {
        assert_eq!(super::part01(_example()), 0);
    }
    #[test]
    fn part02() {
        assert_eq!(super::part02(_example()), 0);
    }
}
// AAA = (BBB, BBB)
fn parse_node(s: &str) -> (&str, (&str, &str)) {
    (&s[0..3], (&s[7..10], &s[12..15]))
}

fn count_it(hm: &HashMap<&str, (&str, &str)>, i: &str, p: &str) -> usize {
    let mut pos = p;
    let mut n = 0;
    while !pos.ends_with("Z") {
        let (l, r) = hm.get(pos).unwrap();
        match i.as_bytes()[n % i.len()] {
            b'L' => pos = l,
            b'R' => pos = r,
            _ => panic!("??"),
        }
        n += 1;
    }
    n
}
pub fn part01(data: String) -> usize {
    let network: Vec<&str> = data.lines().collect();
    let instr = network[0];
    let hm: HashMap<&str, (&str, &str)> =
        HashMap::from_iter(network[2..].iter().map(|q| parse_node(q)));
    count_it(&hm, instr, "AAA")
}

pub fn part02(data: String) -> usize {
    let network: Vec<&str> = data.lines().collect();
    let instr = network[0];
    let hm: HashMap<&str, (&str, &str)> =
        HashMap::from_iter(network[2..].iter().map(|q| parse_node(q)));
    hm.keys()
        .filter(|k| k.ends_with("A"))
        .map(|x| count_it(&hm, instr, x))
        .fold(1, |acc, x| lcm(acc, x))
}
